### PR TITLE
stats: Plumb through symbol tables everywhere. Pulled from #6161.

### LINKED ISSUE #4196: Per listener and per cluster memory overhead is too high
As a result of https://github.com/istio/istio/issues/7912, I've been running bunch of experiments, and it looks that the per listener and per cluster memory overhead is a bit too high...

(numbers with `--concurrency 1 --disable-hot-restart`, include improvements from #4190)

There is ~18KB of memory allocated per each TCP listener (no TLS), ~6KB (~35%) of which is allocated for 5 listener stats (assuming the same `stat_prefix` is used for all TCP stats). There is also ~1450 bytes of overhead per thread/worker for each TCP listener, ~900 bytes of which come from stats.

There is ~63kB of memory allocated per each HTTP listener (1 route, no TLS), ~18kB (~30%) of which is allocated for 13 listener stats (assuming the same `stat_prefix` is used for all HTTP stats). There is also ~1350 bytes of overhead per thread/worker for each HTTP listener, ~900 bytes of which come from stats.

There is ~72kB of memory allocated per each static cluster (no TLS, no health-check), ~58kB (~80%) of which is allocated for 72 stats. There is also ~1850 bytes of overhead per thread/worker for each static cluster.

That means that for each listener and cluster pair, there is ~90kB (TCP) or ~135kB (HTTP) of memory allocated, majority of which is allocated for stats (roughly ~800 bytes per counter/gauge and ~3000 bytes per histogram). Those numbers get even worse when we add TLS to the mix, since it adds 8 more stats for each.

I've noticed that @jmarantz and @ambuc are already working on optimizing stats memory usage in #3585, so I won't be duplicating their work, which leaves a few more drastic options, that potentially yield much bigger savings:
1. add option to disable per listener & per cluster stats (either in config or completely omit them at build-time with `#ifdef`), this is pretty trivial and a huge win, assuming those stats are not consumed,
2. delay cluster initialization until first use, though this might get tricky with EDS and HDS,
3. request CDS from control plane on first use (but that probably wouldn't yield any improvements over 2, would require much bigger changes, and doesn't fit well into eventual consistency mantra).

Thoughts?

cc @htuch @louiscryan @costinm @lizan

### PR DESCRIPTION
**Pull request by jmarantz**
Description: This PR is intended to be a large but trivial and safe refactor, to plumb SymbolTable wherever it will be needed in #6161 and ultimately #4980 . Those PRs have some complexity and areas to watch for performance and contentions, so it will be pretty hard to review them if they are also very large. This is ultimately trying to resolve #4196 .
Risk Level: low - pure refactor
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/common/stats/isolated_store_impl.cc [hunk 1]
```diff
@@ -6,14 +6,24 @@
...
 #include "common/stats/utility.h"
 
 namespace Envoy {
 namespace Stats {
 
 IsolatedStoreImpl::IsolatedStoreImpl()
-    : counters_([this](const std::string& name) -> CounterSharedPtr {
+    : IsolatedStoreImpl(std::make_unique<FakeSymbolTableImpl>()) {}
+
+IsolatedStoreImpl::IsolatedStoreImpl(std::unique_ptr<SymbolTable> symbol_table)
```
Comment (mattklein123) [inline source/common/stats/isolated_store_impl.cc]: nit: Pass as r-value (`&&`)

File: include/envoy/api/api.h [hunk 2]
```diff
@@ -47,6 +47,11 @@ class Api {
    * @return a reference to the TimeSource
    */
   virtual TimeSource& timeSource() PURE;
+
+  /**
+   * @return a reference to the Stats::Store
+   */
+  virtual Stats::Store& statsStore() PURE;
```
Comment (mattklein123) [inline include/envoy/api/api.h]: Q: Why do we need/want stats store as part of the API? This is a little different from the other API objects in that it's a constructed singleton? Mostly just looking for clarification for my own understanding.

Comment (mattklein123) [reply include/envoy/api/api.h]: Well, we already have the `FactoryContext` etc. which we pass to filters for this purpose, and I'm not sure we want to pass the entire `Store` to most extensions, vs. a `Scope`, etc. I think my feeling is probably to take this out and have extensions obtain these type of potentially scoped singletons from the factory context they are passed? WDYT?

Comment (jmarantz) [reply include/envoy/api/api.h]: OK, sgtm. I was thinking we wanted to funnel everything through API at some point but that's kind of speculative. Removed.

File: include/envoy/server/instance.h [hunk 3]
```diff
@@ -209,6 +209,11 @@ class Instance {
    */
   virtual TimeSource& timeSource() PURE;
 
+  /**
+   * @return the statistics symbol table.
+   */
+  virtual Stats::SymbolTable& symbolTable() PURE;
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply include/envoy/server/instance.h]: Fair point; we don't need that. And in fact I can imagine in the future theoretically wanting to use a separate SymbolTable for things other than stats, with no need to comingle them. Removed.

**Pull request merged**